Implement TypeScript namespace support

Wolvereness opened the pull request:
| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #8244, fixes #10038
| Patch: Bug Fix?          | 
| Major: Breaking Change?  | Non-breaking
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2038
| Any Dependency Changes?  |
| License                  | MIT

This adds *basic* namespace support for TypeScript in babel. Currently, it's an outright error, so this only adds new functionality where a project would otherwise outright fail to run through babel. I recommend adding some type of *feature gate* for this functionality, until it has been used at-large for some period of time.

The implementation may be best explained by comparing the [canonical example](https://www.typescriptlang.org/docs/handbook/namespaces.html) that is added as a [test case](https://github.com/babel/babel/tree/28882ced08374d794b36901893853d821d606630/packages/babel-plugin-transform-typescript/test/fixtures/namespace/canonical) (*[in](https://github.com/babel/babel/blob/28882ced08374d794b36901893853d821d606630/packages/babel-plugin-transform-typescript/test/fixtures/namespace/canonical/input.mjs) [out](https://github.com/babel/babel/blob/28882ced08374d794b36901893853d821d606630/packages/babel-plugin-transform-typescript/test/fixtures/namespace/canonical/output.mjs)*).

Three known limitations (ordered by increasing complexity of respective solution):

* ~Sharing the name of the directly enclosing namespace. This requires using a unique identifier as the parameter, which subsequently limits how a project using babel can interact with their namespaces. Making a dynamic solution, one that uses a unique identifier only when necessary, requires a similar solution to the next item. *There is test coverage for this.*~

* Exporting mutable members from inside a namespace. This requires a much more comprehensive depth-first visitor that does extensive scope checking, or an independent state machine to track transformed namespace nodes. *There is test coverage for this.*

* Scope sharing. In image form: ![Scope Sharing](https://images-ext-1.discordapp.net/external/7ogd5EHjOliyTaK12VDrCEbkxVWJJMmYzo8Gh4taLfE/https/media.discordapp.net/attachments/499684844905693184/558651767018094621/unknown.png) [Link](https://www.typescriptlang.org/play/index.html#src=class%20A%20%7B%20%7D%0D%0Anamespace%20A%20%7B%0D%0A%20%20export%20const%20B%20%3D%201%3B%0D%0A%7D%0D%0Anamespace%20A%20%7B%0D%0A%20%20function%20doThing()%20%7B%0D%0A%20%20%20%20console.log(B)%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A). This would require a greater complexity than the mutable member limitation. In addition, a fully correct version requires having a full type-model spanning every file, effectively **impossible** using babel's model.

amaranth commented:
This doesn't correctly handle enums inside namespaces. You end up with invalid syntax.

Input:
```typescript
export namespace A {
    export enum B {
        A = 1,
        B = 2,
        C = 3,
    }
}
```

Output:
```javascript
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.A = void 0;
var A;
exports.A = A;

(function (A) {
  export var B;

  (function (B) {
    B[B["A"] = 1] = "A";
    B[B["B"] = 2] = "B";
    B[B["C"] = 3] = "C";
  })(B || (B = {}));
})(A || (exports.A = A = {}));
```

Config:
```json
{
    "presets": [
        "@babel/env",
        "@babel/typescript"
    ],
    "plugins": [
        "@babel/proposal-class-properties",
        "@babel/proposal-object-rest-spread"
    ]
}
```

Wolvereness commented:
I addressed the issue that @amaranth brought up.

[The incremental change](https://github.com/babel/babel/commit/40e72c4279b3df7550aecfa3994f501a8102a543) [and a minor name change](https://github.com/babel/babel/commit/4df55f0146c058d682aa65f92855d1b9cbf6165c).

I squashed the commits, rebased to current master, and pushed it to this PR.

Wolvereness commented:
[Fixed another issue regarding multiple exports](https://github.com/velocityorg/babel/commit/2fad9d990aa35866b59e546b0c16e836838e672d) [and refactored memberExpression](https://github.com/velocityorg/babel/commit/b9cb021f7440ff255f4b6dc59951f2f98f1eb3c7).

ChuckJonas commented:
Would love to see this merged!  I have a couple projects relying on this feature.  I've already built this branch and linked it to my projects to confirm it work for my use cases.

Wolvereness commented:
Could this get a bit of feedback for the status? I know @martpie @TrySound and @hzoo were giving feedback on the issue and may have some insight for this PR.

nicolo-ribaudo commented on packages/babel-plugin-transform-typescript/src/enum.js:
```diff
@@ -24,7 +24,9 @@ export default function transpileEnum(path, t) {
         const isGlobal = t.isProgram(path.parent); // && !path.parent.body.some(t.isModuleDeclaration);
-        path.replaceWith(makeVar(node.id, t, isGlobal ? "var" : "let"));
+        path.scope.registerDeclaration(
+          path.replaceWith(makeVar(node.id, t, isGlobal ? "var" : "let"))[0],
```
Could you fix this in a separate PR, since it is an actual bugfix and thus less controversial? Probably all the `makeVar` calls need this fix.

Wolvereness replied:
This is the only usage of the call.

Wolvereness replied:
Additional reading: https://github.com/babel/babel/blob/9915c4f3b781483c18e118b6c5bfa4109453126b/packages/babel-plugin-transform-typescript/src/index.js#L70-L75

Wolvereness replied:
#10034

ChuckJonas commented:
@nicolo-ribaudo namespaces ARE NOT deprecated (or is there any plan to).

https://github.com/microsoft/TypeScript/issues/30994

I also believe the `ts-lint` rule is referring to the old legacy global namespace (I could be wrong).   There are many valid reasons to use `namespaces` for code organization.

Really the only reason NOT to use them at this point is:

A: if you are concerned with only writing ts code that conforms to ES6  (but with types)
B: If you want to use babel

By saying "Were not going to support this thing because it's basically legacy", you are essentially making that decision for the typescript team.  This also causes fragmentation in the community and make the build process even more fragile/confusing.

If you're not going to support it, because it's too hard or complex to implement, I understand that.  But please stop telling people that it's a "legacy feature".

Wolvereness commented:
> 1. As you pointed out, there is a big number of caveats. Even if it is not always possible, we try to match the spec closely: that's why, for example, we are testing our parser agains the test262 test suite. Implementing namespaces exactly as TS does is currently impossible because of the cross-file merging but, as you pointed out, even ignoring it leaves an high implementation complexity.

I believe that the caveats would be uncommon as a limitation for a user. I included them so they would be documented and clearly benign against a common-case (the last of which I'd even put a large gamble a special case could be constructed to trip up the official compiler if types aren't declared fully). As implemented, this should satisfy the overwhelming majority of how namespaces are used. *If there is a caveat not listed, then it needs to be brought to my attention.*

I assumed Babel isn't taking an ideological stance, and more one from how developer time is diverted. With this PR, the time has been put in, and namespaces can be supported by Babel on a best-effort basis, similar to the rest of the typescript implementation.

> 2. TypeScript namespaces are pretty much considered derpecated: as TypeScript's PM [wrote](https://github.com/babel/babel/issues/8244#issuecomment-408187141), there are better alternatives for namespaces. Also, there is [a reccomended rule](https://github.com/palantir/tslint/blob/3efdfe57b7c0eb1b21e21cec128c993cf20da438/src/configs/recommended.ts#L85) in TSLint to disallow them.

https://github.com/microsoft/TypeScript/issues/30994

> Possible approaches I can think of are:
> 
> * Keeps things as-is;
> * Implement namespaces as it's done in this PR, even if it has all those known limitations;
> * Only accept type-only namespaces, which can be safely removed without affecting any other code.

Type-only namespaces are (should-be) marked with declare, and are already removed when marked as such. https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-typescript/test/fixtures/declarations/erased

> This PR also fixes an actual bug, and it would be ideal if you (or anyone else) could open a separate PR which only fixes it, since it is far less controversial than adding namespaces support.

I'm not sure this is an actual bug. I asked about it on the slack and was met with radio silence. The behavior of whether or not something is in the scope doesn't seem to matter while the TS plugin runs (it doesn't use scope), and similarly there's a different expectation that an enum would get handled as a class would *somewhere else*.

Or to rephrase, I think the only *observable* behavior of it relies on this PR. Also, my interest is rather specific to supporting namespaces, so me putting in the effort of another PR is contingent on the outlook for this PR.

nicolo-ribaudo commented:
Thanks for linking microsoft/TypeScript#30994. I definetly had a wrong understanding of the current `namespace` status.

---

I was playing with this PR on a [repl which supports TS](https://deploy-preview-2019--babel.netlify.com/repl#?build=10686&filename=repl.ts) and noticed this:

```ts
// Throws
namespace Bar {
  export let a = 2;
  a = 3;
}

// Does not throw
namespace Foo {
  export function y() {}
  y = 3;
}
```

is it expected?

Wolvereness commented:
It's not valid typescript. That's a type-error, something that babel shouldn't be expected to handle.
![image](https://user-images.githubusercontent.com/983111/57642012-044c1400-757c-11e9-9a55-a49fc368c463.png)

Wolvereness commented on packages/babel-plugin-transform-typescript/src/namespace.js:
```diff
@@ -0,0 +1,205 @@
+      default:
+        continue;
+      case "ExportNamedDeclaration":
+    }
```
Nodes that are exporting (`ExportNamedDeclaration`) need to get transpiled. The rest/other switch of the method deals with those cases, which are effectively the *primary case* (what really matters) for the entire function. The other cases are simplistically tracking the usages of names, effectively incidental.

Alternatives:

* I can move the entire switch statement up into that case and drop the default-case. I didn't do this because it felt unnecessary.

* I can add a comment(s) explaining that the case that matters most is just the `export` case, and anything else just continues in the loop.

* I can the second switch into a function, to make it more concise.  I didn't do this because it's a lot of parameter sharing, especially for a *recursive* function. However, I would prefer this over *nesting* switches (first alternative).

nicolo-ribaudo replied:
> I can add a comment(s) explaining that the case that matters most is just the export case, and anything else just continues in the loop.

:+1:

Wolvereness commented:
> > Sharing the name of the directly enclosing namespace. This requires using a unique identifier as the parameter, which subsequently limits how a project using babel can interact with their namespaces.
> 
> I don't fully understand this restriction. Wouldn't it be possible to transpile namespaces like this, to avioid the naming conflict?

Solving this selectively makes it two-pass. One for figuring out if we should use a unique identifier, and the second for the transforms.

We **could** have it **always** use an assumed-unique identifier. This creates its own caveat to be at least slightly divergent from an edge case in typescript if something reassigns the identifier representing the namespace itself post-initialization. It also creates another caveat; (actual question from me, not hypothetical) does the process for generating a unique identifier consider that it will be for a sub-scope? Those aren't in the `scope` for the parent, and deciding something (provably) unique requires it to have been inspected already.

nicolo-ribaudo commented:
Yes, it generates an identifier which is unique in the whole file.
We only need to generate the UUID for internal usage in the namespace function (and we can always to that); the external identifier shall remain the same. 

> This creates its own caveat to be at least slightly divergent from an edge case in typescript if something reassigns the identifier representing the namespace itself post-initialization.

Can you share an example?

Wolvereness commented:
> > This creates its own caveat to be at least slightly divergent from an edge case in typescript if something reassigns the identifier representing the namespace itself post-initialization.
> 
> Can you share an example?

I was setting it up and realized it's actually going to hit a type error. I'm going to assume now what I was thinking of isn't really possible, but I'll do a mental proof sometime this weekend to assure myself, before I write the UUID change.

Wolvereness commented:
> It should generate an ID which is unique in the whole file. I'm not 100% sure though.

I actually found a bug testing this PR, not that it's particularly relevant. `enum` values are not unique, but everything else appears to be.

Wolvereness commented:
Namespace name clashes are no longer an issue.

nicolo-ribaudo commented:
Hi all :wave: We talked about this PR during the last meeting ([notes](https://github.com/babel/notes/blob/master/2019/05/21.md)).

The outcome was:
- As suggested in the PR description, namespaces should be enabled behind a flag (`allowNamespaces`). We will enable it by default in the future.
- Could someone document the limitations at https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats? :pray: :slightly_smiling_face:

Wolvereness commented:
I'm very glad to hear that!

* Is there an example of a feature flag that I could mimic, or will someone else be working on that?

* The meeting notes mention making sure the exceptions are user friendly. Should I add the URL to the documentation for the *non-const export* exception? I would assume an exception for encountering a namespace outside of the feature flag may also have the URL.

* I'm probably a good candidate to write the PR for https://github.com/babel/website/blob/master/docs/plugin-transform-typescript.md. I can work on that later in the week.

danez commented:
> * Is there an example of a feature flag that I could mimic, or will someone else be working on that?

It is basically the same as this options: https://babeljs.io/docs/en/babel-plugin-transform-typescript#options (`jsxPragma` for example)
If you have time, it would be nice if you could add an option. Otherwise it's also fine and we can do it.

> * The meeting notes mention making sure the exceptions are user friendly. Should I add the URL to the documentation for the _non-const export_ exception? I would assume an exception for encountering a namespace outside of the feature flag may also have the URL.

Adding URLs sounds good. Our concern was that users who run into this limitations would think it is a  bug and report it and a good error message could clarify that these cases are unsupported.

Thanks for all the effort.

nicolo-ribaudo commented on packages/babel-plugin-transform-typescript/src/namespace.js:
```diff
@@ -0,0 +1,177 @@
+    assignment,
+  );
+
+  return t.expressionStatement(
```
And here
```js
return template.statement.ast`
  (function (${name}) {
    ${namespaceTopLevel}
  })(${id} || (${id} = ${fallthroughValue}));
`;
```

Wolvereness replied:
bb95958dd0c5fbf0ff41293dbb10dd1097d6e783

Wolvereness commented on packages/babel-plugin-transform-typescript/src/namespace.js:
```diff
@@ -0,0 +1,177 @@
+      path.replaceWith(getDeclaration(t, name));
+      path.scope.registerDeclaration(path.parentPath);
+    } else {
+      path.parentPath.replaceWith(value);
```
It's not actually valid typescript, but for some reason typescript wont fuss if the namespace itself is empty.

So, I'm a bit confused, what behavior should we be *testing*?

amaranth replied:
I believe TypeScript ignores this because it doesn't emit anything for an empty namespace. That logic likely accidentally makes it ignore the type error as well.

nicolo-ribaudo replied:
Oh ok; maybe typescript accepts it because it is considered as a type-only namespace?

Wolvereness commented on packages/babel-plugin-transform-typescript/src/namespace.js:
```diff
@@ -0,0 +1,177 @@
+      case "TSEnumDeclaration":
+      case "FunctionDeclaration":
+      case "ClassDeclaration":
+        names[subNode.id.name] = true;
```
I was going to argue about my preferences, but then I realized that this is actually introduces random bugs (bc65315f8f8db47e158e5ffdd5d53fa9e493bbb4).

Wolvereness commented on packages/babel-plugin-transform-typescript/src/namespace.js:
```diff
@@ -0,0 +1,177 @@
+      "||",
+      getMemberExpression(t, parentExportName, realName),
+      exportAssignment,
+    );
```
bb95958dd0c5fbf0ff41293dbb10dd1097d6e783

Wolvereness commented:
Experimental flag added: 4d447b3b6db75a182416823e6ec4b5bcf30ce693

Better error message for non-const: 6d5df4328c10c9e7a24a6d7a13708bd123579c42

Wolvereness commented:
Rebased and squashed.

nicolo-ribaudo commented:
Could you please also put the fix for #10033 in  a separate PR? :pray: It can go in the next patch release, while this PR needs to wait for a minor.

Wolvereness commented:
> Could you please also put the fix for #10033 in a separate PR? 🙏 It can go in the next patch release, while this PR needs to wait for a minor.

#10034

nicolo-ribaudo commented:
Does thir PR fix https://github.com/babel/babel/issues/10038?

Wolvereness commented:
> Does thir PR fix #10038?

Added as a fixture (so, not a full-babel transform), and this is the output (I'm not sure if it's correct):

```javascript
let src;

(function (_src) {
  let ns1;

  (function (_ns) {
    class foo {}

    _ns.foo = foo;
  })(ns1 || (ns1 = _src.ns1 || (_src.ns1 = {})));

  let ns2;

  (function (_ns2) {
    class foo {}

    _ns2.foo = foo;
  })(ns2 || (ns2 = _src.ns2 || (_src.ns2 = {})));
})(src || (src = {}));
```

nicolo-ribaudo commented:
Yeah, it seems correct. In the issue it didn't even produce an output :stuck_out_tongue: 	

> so, not a full-babel transform

What do you mean?

Wolvereness commented:
> > so, not a full-babel transform
> 
> What do you mean?

Fixtures are unit-tests. An end-user would rarely need *only* the plugin that the fixtures test. For example, [the typescript-supporting repl](https://deploy-preview-2019--babel.netlify.com/repl#?babili=false&browsers=&build=10686&builtIns=false&spec=false&loose=false&code_lz=PTAEBcE8AcFMGcDGAnAltcAoAtgewCYCuANrKPMoqAN6aj2iwAe0uy4oAdgIbYLTdEZTvACMNOgynNW7UImLd48UADNcuCVO0MAYqIBc5cGk4BzANySdAX2ug70lmw48-8AUK7wATFp2MznIKSirqmrQB2vpG8Cao5lZRjgx2NkA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&filename=repl.ts&lineWrap=true&presets=es2015%2Creact%2Cstage-2%2Ctypescript&prettier=false&targets=&version=7.4.3%2Bpr.9785&externalPlugins=) has a bit more output. It also changes the use of underscores.

```javascript
"use strict";

function _instanceof(left, right) { if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) { return right[Symbol.hasInstance](left); } else { return left instanceof right; } }

function _classCallCheck(instance, Constructor) { if (!_instanceof(instance, Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// typescript
var src;

(function (src) {
  var ns1;

  (function (ns1) {
    var foo = function foo() {
      _classCallCheck(this, foo);
    };

    ns1.foo = foo;
  })(src.ns1 = ns1 || (ns1 = {}));

  var ns2;

  (function (ns2) {
    var foo = function foo() {
      _classCallCheck(this, foo);
    };

    ns2.foo = foo;
  })(src.ns2 = ns2 || (ns2 = {}));
})(src || (src = {}));
```

nicolo-ribaudo commented:
Oh I see; the repl output is different because there are the es2015 and stage-2 presets enabled

nicolo-ribaudo commented:
Merged at 0a988143290da1208475aed84c0429591df7d0f8
